pulser/simulation/simresults.py

```diff
@@ -14,63 +14,320 @@
+        return [self._results[t_index][b] for b in bitstrings]
+
+    def _standard_dev(self, op: qutip.Qobj) -> ArrayLike:
+        """Returns the square root of the variance of operator op."""
+        return cast(ArrayLike,
+                    np.sqrt(qutip.variance(op, self.states) / self.N_measures)
+                    )
+
+    def _get_error_bars(self, op: qutip.Qobj) -> Tuple[ArrayLike, ArrayLike]:
+        moy = self.expect([op])[0]
+        st = self._standard_dev(op)
+        return moy, st
+
+    def plot(self, op: qutip.Qobj, fmt: str = '.',
+             label: str = '', error_bars: bool = True) -> None:
+        """Plots the expectation value of a given operator op.
+
+        Note: The observable must be diagonal.
+
+        Args:
+            op (qutip.Qobj): Operator whose expectation value is wanted.
+            error_bars (bool): Choose to display error bars.
+            fmt (str): Curve plot format.
+            label (str): y-Axis label.
+        """
+        if not isdiagonal(op):
+            raise ValueError(f"Observable {op} is non-diagonal.")
+        if error_bars:
+            moy, st = self._get_error_bars(op)
+            plt.errorbar(self.sim_times, moy, st, fmt=fmt, lw=1, capsize=3,
+                         label=label)
+        else:
+            plt.plot(self.sim_times, self.expect([op])[0], fmt,
+                     label=label)
+        plt.xlabel('Time (µs)')
+        plt.ylabel('Expectation value')
+
+
+class CleanResults(SimulationResults):
+    """Results of an ideal simulation run of a pulse sequence.
+
+    Contains methods for studying the states and extracting useful information
+    from them.
+    """
+
+    def __init__(self, run_output: list[qutip.Qobj],
+                 dim: int, size: int, basis_name: str,
+                 sim_times: np.ndarray, meas_basis: str) -> None:
+        """Initializes a new CleanResults instance.
+
+        Args:
+            run_output (list of qutip.Qobj): List of `qutip.Qobj` corresponding
+                to the states at each time step after the evolution has been
+                simulated.
+            dim (int): The dimension of the local space of each atom (2 or 3).
+            size (int): The number of atoms in the register.
+            basis_name (str): The basis indicating the addressed atoms after
+                the pulse sequence ('ground-rydberg', 'digital' or 'all').
+            sim_times (list): Times at which Simulation object returned the
+                results.
+            meas_basis (str): The basis in which a sampling measurement
+                is desired.
+        """
+        super().__init__(dim, size, basis_name, sim_times)
         if meas_basis:
             if meas_basis not in {'ground-rydberg', 'digital'}:
                 raise ValueError(
                     "`meas_basis` must be 'ground-rydberg' or 'digital'."
                     )
         self._meas_basis = meas_basis
+        self._results = run_output
```

Comment:
Is there a reason to rename `self._states` to `self._results` on `CleanResults`? I guess it doesn't hurt, just wondering

Reply:
Not really, but I thought it'd be consistent with `NoisyResults` this way

Reply:
That's what I thought. It's okay like this, no need to change it